powerauth-restful-security-spring-annotation/src/main/java/io/getlime/security/powerauth/rest/api/spring/annotation/PowerAuthAnnotationInterceptor.java

```diff
@@ -147,4 +154,53 @@ public boolean preHandle(HttpServletRequest request, HttpServletResponse respons
         return Object.class;
     }
 
+    /**
+     * The method substitutes placeholders (marked via "${placeholder}") in resourceID attribute value by
+     * the actual parameters of the handler method. The implementation takes into account all method parameters
+     * that are annotated via @RequestParam or @PathVariable annotations and extracts values from the request
+     * parameter map.
+     *
+     * @param resourceId Resource ID with possible placeholders.
+     * @param request HttpServletRequest for the current execution.
+     * @param handlerMethod Handler method that is responsible for the request processing.
+     * @return Resource ID with substituted placeholders.
+     */
+    @SuppressWarnings("unchecked")
+    private String expandResourceId(String resourceId, HttpServletRequest request, HandlerMethod handlerMethod) {
+        // Get method parameters that could be replaced in the context of resource ID
+        final Map<String, String> parameters = new TreeMap<>();
+        final MethodParameter[] methodParameters = handlerMethod.getMethodParameters();
+        for (MethodParameter mp : methodParameters) {
+            // Handle parameters annotated by @RequestParam annotation.
+            // These are stored in the servlet request parameter map.
+            final RequestParam requestParam = mp.getParameterAnnotation(RequestParam.class);
+            if (requestParam != null) {
+                final String name = requestParam.name();
+                final String value = request.getParameter(name);
+                if (value != null) {
+                    parameters.put(name, value);
+                }
+            } else {
+                // Handle parameters annotated by @PathVariable annotation.
+                // These are stored by Spring in the servlet request attributes map, under a special
+                // URI_TEMPLATE_VARIABLES_ATTRIBUTE key that contains Map<String, String> with path
+                // variable mapping.
```

Comment:
Perhaps we should add a comment that in case of a parameter name conflict the path variables take precendence over request parameters. Or alternatively check for variable name conflicts.